pinot-common/src/main/java/org/apache/pinot/common/utils/PinotThrottledLogger.java

```diff
@@ -0,0 +1,158 @@
+ * PinotThrottledLogger throttled = new PinotThrottledLogger(logger, ingestionConfig, tableName);
+ *
+ * try {
+ *   // some operation
+ * } catch (Exception e) {
+ *   throttled.warn("Operation failed for record: " + record, e);
+ * }
+ * </pre>
+ *
+ * <p><b>Logging Behavior:</b>
+ * <ul>
+ *   <li><b>Rate limit disabled (≤ 0):</b> All exceptions logged at DEBUG level (backward compatible default)</li>
+ *   <li><b>Within rate limit (> 0, quota available):</b> Logs at WARN/ERROR level with suppression counts</li>
+ *   <li><b>Rate limit exceeded (> 0, quota exhausted):</b> Logs at DEBUG level while tracking suppression counts</li>
+ * </ul>
+ * This ensures no exception information is lost - all exceptions are logged at either WARN/ERROR or DEBUG level.
+ *
+ * <p><b>Example Output:</b>
+ * <pre>
+ * WARN  [MyClass] Operation failed for record: {id=1}
+ * java.lang.NumberFormatException: For input string: "abc"
+ *     at java.lang.NumberFormatException.forInputString(...)
+ *
+ * [... 4 more similar logs within 1 minute ...]
+ *
+ * [After rate limit window passes and 10,001st exception occurs]
+ * WARN  [MyClass] ... Suppressed 9995 occurrences of NumberFormatException ...
+ * WARN  [MyClass] Operation failed for record: {id=10001}
+ * java.lang.NumberFormatException: For input string: "xyz"
+ * </pre>
+ *
+ * <p>Meanwhile, if a different exception type occurs (e.g., ConnectException), it logs immediately using its own
+ * independent rate limiter, ensuring critical errors are never starved by high-frequency errors.
+ *
+ * @see org.apache.pinot.spi.config.table.ingestion.IngestionConfig#getIngestionExceptionLogRateLimitPerMin()
+ */
+public class PinotThrottledLogger {
+  private final Logger _delegate;
+
+  private final Map<Class<?>, ExceptionState> _stateMap = new HashMap<>();
+  private final double _permitsPerSecond;
+  private final String _tableName;
+
+  public PinotThrottledLogger(Logger delegate, @Nullable IngestionConfig ingestionConfig, @Nullable String tableName) {
+    this(delegate, getPermitsPerSecond(ingestionConfig), tableName);
+  }
+
+  public PinotThrottledLogger(Logger delegate, double permitsPerSecond) {
+    this(delegate, permitsPerSecond, null);
+  }
+
+  public PinotThrottledLogger(Logger delegate, double permitsPerSecond, @Nullable String tableName) {
+    _delegate = delegate;
+    _permitsPerSecond = permitsPerSecond;
+    _tableName = tableName;
+  }
+
+  private static double getPermitsPerSecond(IngestionConfig ingestionConfig) {
+    return Optional.ofNullable(ingestionConfig).orElse(new IngestionConfig())
+        .getIngestionExceptionLogRateLimitPerMin() / 60.0;
+  }
+
+  public void warn(String msg, Throwable t) {
+    logWithRateLimit(msg, t, _delegate::warn);
+  }
+
+  public void error(String msg, Throwable t) {
+    logWithRateLimit(msg, t, _delegate::error);
+  }
+
+  private void logWithRateLimit(String msg, Throwable t, BiConsumer<String, Throwable> consumer) {
+    if (_permitsPerSecond <= 0) {
```

Comment:
Consider modeling this as a separate `DebugOnlyLogger`. We can also have a `UnthrottledLogger` if necessary